tests/Acoustics.Test/AudioAnalysisTools/DSP/KmeansClusteringTests.cs

```diff
@@ -0,0 +1,226 @@
+
+                //Draw clusters
+                var clusterImage = ImageTools.DrawMatrixWithoutNormalisation(mergedCentroidMatrix);
+                clusterImage.RotateFlip(RotateFlipType.Rotate270FlipNone);
+
+                var outputClusteringImage = Path.Combine(outputDir.FullName, "ClustersWithGrid" + i.ToString() + ".bmp");
+                FrequencyScale.DrawFrequencyLinesOnImage((Bitmap)clusterImage, freqScale, includeLabels: false);
+                clusterImage.Save(outputClusteringImage);
+            }
+
+            //+++++++++++++++++++++++++++++++++++++++++++++++++++++Processing and generating features for the target spectrogram
+            var recording2Path = PathHelper.ResolveAsset("Recordings", "BAC2_20071008-085040.wav");
+            var recording2 = new AudioRecording(recording2Path);
+            var sonogram2 = new SpectrogramStandard(sonoConfig, recording2.WavReader);
+            var targetSpec = sonogram2.Data;
+
+            //Do RMS normalization
+            sonogram2.Data = PcaWhitening.RmsNormalization(sonogram2.Data);
+
+            //NOISE REDUCTION
+            sonogram2.Data = PcaWhitening.NoiseReduction(sonogram2.Data);
+
+            //extracting sequential patches from the target spectrogram
+            List<double[,]> allSubmatrices2 = PatchSampling.GetFreqBandMatrices(sonogram2.Data, noOfFreqBand);
+            double[][,] matrices2 = allSubmatrices2.ToArray();
+            List<double[,]> allSequentialPatchMatrix = new List<double[,]>();
+            for (int i = 0; i < matrices2.GetLength(0); i++)
+            {
+                int rows = matrices2[i].GetLength(0);
+                int cols = matrices2[i].GetLength(1);
+                var sequentialPatches = PatchSampling.GetPatches(matrices2[i], patchWidth, patchHeight, (rows / patchHeight) * (cols / patchWidth), "sequential");
+                allSequentialPatchMatrix.Add(sequentialPatches.ToMatrix());
+            }
+
+            List<double[,]> convertedSpec = new List<double[,]>();
+            int columnPerFreqBand = sonogram2.Data.GetLength(1) / noOfFreqBand;
+            for (int i = 0; i < allSequentialPatchMatrix.Count; i++)
+            {
+                double[,] reconstructedSpec2 = KmeansClustering.ReconstructSpectrogram(allSequentialPatchMatrix.ToArray()[i], allClusteringOutput.ToArray()[i]);
+                convertedSpec.Add(PatchSampling.ConvertPatches(reconstructedSpec2, patchWidth, patchHeight, columnPerFreqBand));
+            }
+
+            sonogram2.Data = PatchSampling.ConcatFreqBandMatrices(convertedSpec);
+
+            // DO DRAW SPECTROGRAM
+            var respecImage = sonogram2.GetImageFullyAnnotated(sonogram2.GetImage(), "RECONSTRUCTEDSPECTROGRAM: " + freqScale.ScaleType.ToString(), freqScale.GridLineLocations);
+            respecImage.Save(outputImagePath, ImageFormat.Png);
+
+            // DO UNIT TESTING
+            Assert.AreEqual(targetSpec.GetLength(0), sonogram2.Data.GetLength(0));
```

Comment:
It's great these tests pass, however you should not they're very brittle.

You don't have to change them now, but consider how any small change in the code (e.g. how we compute FFTs) could cause these tests to fail. How would you add a tolerance?

Reply:
You are right. I will work on the unit tests.